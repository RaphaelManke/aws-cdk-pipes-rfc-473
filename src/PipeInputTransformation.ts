
enum reservedVariables {
  PIPES_ARN = '<aws.pipes.pipe-arn>',
  PIPES_NAME = '<aws.pipes.pipe-name>',
  PIPES_TARGET_ARN = '<aws.pipes.target-arn>',
  PIPE_EVENT_INGESTION_TIME = '<aws.pipes.event.ingestion-time>',
  PIPE_EVENT = '<aws.pipes.event>',
  PIPE_EVENT_JSON = '<aws.pipes.event.json>',
}

type StaticString = string;
// type JsonPath = `<$.${string}>`;
type KeyValue = Record<string, string | reservedVariables>;
type StaticJsonFlat = Record<string, StaticString | KeyValue>;
type InputTransformJson = Record<
string,
StaticString | KeyValue | StaticJsonFlat
>;

type PipeInputTransformationValue = StaticString | InputTransformJson;

export interface IInputTransformationProps {
  inputTemplate: PipeInputTransformationValue;
}

export class PipeInputTransformation {
  static fromJson(inputTemplate: Record<string, any>): PipeInputTransformation {
    return new PipeInputTransformation({ inputTemplate });
  }

  readonly inputTemplate: string;

  constructor(props: IInputTransformationProps) {
    this.inputTemplate = JSON.stringify(props);
  }
}
