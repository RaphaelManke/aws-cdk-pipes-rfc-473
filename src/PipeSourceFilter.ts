export interface IPipeFilterPattern {
  pattern: string;
}

export interface IPipeSourceFilter {
  filters: IPipeFilterPattern[];
}

export class PipeGenericFilterPattern {
  static fromJson(patternObject: Record<string, any>): IPipeFilterPattern {
    return { pattern: JSON.stringify(patternObject) };
  }
}

export class PipeSourceFilter implements IPipeSourceFilter {
  public filters: IPipeFilterPattern[];


  constructor(filter: IPipeFilterPattern[]) {
    this.filters = filter;
  }

  public toString(): string {
    return JSON.stringify(this.filters);
  }
}
