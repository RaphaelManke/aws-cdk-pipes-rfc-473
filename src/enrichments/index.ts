export * from './ApiGatewayEnrichment';
export * from './LambdaEnrichment';
export * from './StepFunctionEnrichment';