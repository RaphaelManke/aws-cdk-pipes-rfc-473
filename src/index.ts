export * from './enrichments';
export * from './Pipe';
export * from './PipeEnrichment';
export * from './PipeInputTransformation';
export * from './PipeSource';
export * from './PipeSourceFilter';
export * from './PipeTarget';
export * from './sources';
export * from './targets';
