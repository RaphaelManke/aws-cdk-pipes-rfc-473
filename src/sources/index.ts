export * from './ActiveMqBrokerSource';
export * from './AmazonMqBrokerSource';
export * from './DynamoDBStreamSource';
export * from './KinesisStreamSource';
export * from './MskSource';
export * from './PipeSourceDeadLetterConfig';
export * from './PipeSourceOnPartialBatchItemFailure';
export * from './PipeSourceStartingPosition';
export * from './RabbitMqBrokerSource';
export * from './SelfMangedKafkaSource';
export * from './SqsSource';