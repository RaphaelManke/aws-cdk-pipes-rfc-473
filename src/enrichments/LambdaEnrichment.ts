import { IRole } from 'aws-cdk-lib/aws-iam';
import { IFunction } from 'aws-cdk-lib/aws-lambda';
import { PipeEnrichment } from '../PipeEnrichment';
import { PipeInputTransformation } from '../PipeInputTransformation';

export class LambdaEnrichment extends PipeEnrichment {
  private lambda : IFunction;

  constructor(lambda: IFunction, props: {inputTransformation?: PipeInputTransformation}) {
    super(lambda.functionArn, { inputTemplate: props.inputTransformation?.inputTemplate });
    this.lambda = lambda;
  }

  grantInvoke(grantee: IRole): void {
    this.lambda.grantInvoke(grantee);
  }
}