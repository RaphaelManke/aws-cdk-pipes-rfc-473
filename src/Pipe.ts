import {
  IResource,
  Lazy,
  Names,
  Resource,
} from 'aws-cdk-lib';
import { IRole, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { Construct } from 'constructs';
import { IPipeEnrichment } from './PipeEnrichment';
import { IPipeSource } from './PipeSource';
import { PipeSourceFilter as IPipeSourceFilter } from './PipeSourceFilter';
import { IPipeTarget } from './PipeTarget';

/**
 * Interface all Pipe constructs must implement
 */
export interface IPipe extends IResource {
  /**
   * The name of the pipe
   *
   * @attribute
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html#cfn-pipes-pipe-name
   */
  readonly pipeName: string;

  /**
   * The ARN of the pipe
   *
   * @attribute
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html#Arn-fn::getatt
   */
  readonly pipeArn: string;

  /**
   * The role used by the pipe
   *
   * @attribute
   */
  readonly pipeRole: IRole;
}

export enum DesiredState {
  RUNNING = 'RUNNING',
  STOPPED = 'STOPPED',
}


export interface IPipeProps {
  /**
   * The source of the pipe
   */
  readonly source: IPipeSource;
  /**
   * The filter pattern for the pipe source
   */
  readonly filter?: IPipeSourceFilter;
  /**
  *
  */
  readonly enrichment?: IPipeEnrichment;
  /**
   * The target of the pipe
   */
  readonly target: IPipeTarget;
  /**
  * Name of the pipe in the AWS console
  *
  * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html#cfn-pipes-pipe-name
  */
  readonly name?: string;
  /**
   * The role used by the pipe which has permissions to read from the source and write to the target.
   * If an enriched target is used, the role also have permissions to call the enriched target.
   * If no role is provided, a role will be created.
   */
  readonly role?: IRole;
  /**
   * A description of the pipe displayed in the AWS console
   *
   * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html#cfn-pipes-pipe-description
   */
  readonly description?: string;
  /**
   * The desired state of the pipe. If the state is set to STOPPED, the pipe will not process events.
   *
   * @link https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_Pipe.html#eventbridge-Type-Pipe-DesiredState
   */
  readonly desiredState?: DesiredState;
  /**
   * `AWS::Pipes::Pipe.Tags`
   *
   * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html#cfn-pipes-pipe-tags
   */
  readonly tags?: {
    [key: string]: string;
  };
}

abstract class PipeBase extends Resource implements IPipe {
  public abstract readonly pipeName: string;
  public abstract readonly pipeArn: string;
  public abstract readonly pipeRole: IRole;
}

export class Pipe extends PipeBase {
  public readonly pipeName: string;
  public readonly pipeArn: string;
  public readonly pipeRole: IRole;

  constructor(scope: Construct, id: string, props: IPipeProps) {
    const pipeName =
      props.name || Lazy.string({ produce: () => Names.uniqueId(this) });
    super(scope, id, { physicalName: pipeName });

    this.pipeRole =
      props.role ||
      new Role(this, 'Role', {
        assumedBy: new ServicePrincipal('pipes.amazonaws.com'),
      });
    this.pipeName = pipeName;

    const sourceParameters = {
      ...props.source.sourceParameters,
      filterCriteria: props.filter,
    };
    if (props.enrichment) {
      props.enrichment.grantInvoke(this.pipeRole);
    }
    props.source.grantRead(this.pipeRole);
    props.target.grantPush(this.pipeRole);

    const resource = new CfnPipe(this, 'Resource', {
      name: props.name,
      source: props.source.sourceArn,
      sourceParameters: sourceParameters,
      target: props.target.targetArn,
      targetParameters: props.target.targetParameters,
      enrichment: props.enrichment?.enrichmentArn,
      roleArn: this.pipeRole.roleArn,
      description: props.description,
      desiredState: props.desiredState,
      enrichmentParameters: props.enrichment?.enrichmentParameters,
      tags: props.tags,
    });
    this.pipeArn = resource.attrArn;
  }
}
