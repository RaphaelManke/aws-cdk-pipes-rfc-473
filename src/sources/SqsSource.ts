import { IRole } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';
import { IQueue } from 'aws-cdk-lib/aws-sqs';
import { PipeSource } from '../PipeSource';
import { IPipeFilterPattern, PipeGenericFilterPattern } from '../PipeSourceFilter';


export class SqsSource extends PipeSource {
  private queue: IQueue;

  constructor(queue: IQueue, props?:CfnPipe.PipeSourceSqsQueueParametersProperty) {
    super(queue.queueArn, { sqsQueueParameters: props });
    this.queue = queue;

  }

  public grantRead(grantee: IRole): void {
    this.queue.grantConsumeMessages(grantee);
  }
}


export interface ISqsAttributes {
  approximateReceiveCount?: string;
  sentTimestamp?: string;
  sequenceNumber?: string;
  messageGroupId?: string;
  senderId?: string;
  messageDeduplicationId?: string;
  approximateFirstReceiveTimestamp?: string;
}

export interface ISqsMessagePipeFilter {
  messageId?: string;
  receiptHandle?: string;
  body?: any;
  attributes?: ISqsAttributes;
  messageAttributes?: any;
  md5OfBody?: string;
}

export class PipeSqsFilterPattern extends PipeGenericFilterPattern {
  static fromSqsMessageAttributes(
    attributes: ISqsMessagePipeFilter,
  ): IPipeFilterPattern {
    const sqsProps = {
      ApproximateReceiveCount: attributes.attributes?.approximateReceiveCount,
      SentTimestamp: attributes.attributes?.sentTimestamp,
      SequenceNumber: attributes.attributes?.sequenceNumber,
      MessageGroupId: attributes.attributes?.messageGroupId,
      SenderId: attributes.attributes?.senderId,
      MessageDeduplicationId: attributes.attributes?.messageDeduplicationId,
      ApproximateFirstReceiveTimestamp:
        attributes.attributes?.approximateFirstReceiveTimestamp,
    };

    return {
      pattern: JSON.stringify({
        ...attributes,
        attributes: attributes.attributes ? sqsProps : undefined,
      }),
    };
  }
}