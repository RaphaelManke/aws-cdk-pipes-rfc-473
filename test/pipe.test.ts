import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { LambdaEnrichment } from '../src/enrichments/LambdaEnrichment';
import { Pipe } from '../src/Pipe';
import { PipeInputTransformation } from '../src/PipeInputTransformation';
import { PipeGenericFilterPattern, PipeSourceFilter } from '../src/PipeSourceFilter';

import { PipeSqsFilterPattern, SqsSource } from '../src/sources/SqsSource';
import { SqsTarget } from '../src/targets/SqsTarget';

describe('Pipe', () => {

  it('Stack does synth', () => {
    const stack = new Stack(new App(), 'test-stack');
    const sourceQueue = new Queue(stack, 'test-source-queue', {});
    const targetQueue = new Queue(stack, 'test-target-queue', {});

    const pipeSourceFilter = new PipeSourceFilter([
      PipeGenericFilterPattern.fromJson({ foo: 'bar' }),
      PipeSqsFilterPattern.fromSqsMessageAttributes({
        md5OfBody: 'string',
        body: { foo: 'bar' },
      }),
    ]);

    const enrichmentLambda = new Function(stack, 'test-lambda', {
      code: Code.fromInline(
        'exports.handler = async (event) => { return event; };',
      ),
      runtime: Runtime.NODEJS_18_X,
      handler: 'index.handler',
    });

    new Pipe(stack, 'test-pipe', {
      description: 'test-pipe',

      source: new SqsSource(sourceQueue, {

      }),
      filter: pipeSourceFilter,

      enrichment: new LambdaEnrichment(enrichmentLambda, {
        inputTransformation: PipeInputTransformation.fromJson({ foo: 'bar' }),
      }),
      target: new SqsTarget(targetQueue),

    });


    const template = Template.fromStack(stack);

    template.hasResource('AWS::Pipes::Pipe', {});
    template.hasResource('AWS::IAM::Role', {});

    const queues = Object.values(template.findResources('AWS::SQS::Queue', {}));

    expect(queues).toHaveLength(2);


    const pipeRolePolicy = Object.values(
      template.findResources('AWS::IAM::Policy', {}),
    )[0];

    expect(pipeRolePolicy.Properties.PolicyDocument.Statement).toHaveLength(3);

    expect(template).toMatchSnapshot();
  });
});