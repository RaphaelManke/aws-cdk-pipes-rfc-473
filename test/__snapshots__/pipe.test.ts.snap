// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipe Stack does synth 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testlambdaC73320F8": Object {
      "DependsOn": Array [
        "testlambdaServiceRole28D9F97A",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = async (event) => { return event; };",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testlambdaServiceRole28D9F97A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testlambdaServiceRole28D9F97A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testpipeEBFCD55C": Object {
      "Properties": Object {
        "Description": "test-pipe",
        "Enrichment": Object {
          "Fn::GetAtt": Array [
            "testlambdaC73320F8",
            "Arn",
          ],
        },
        "EnrichmentParameters": Object {
          "InputTemplate": "{\\"foo\\":\\"bar\\"}",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "testpipeRole4CA84261",
            "Arn",
          ],
        },
        "Source": Object {
          "Fn::GetAtt": Array [
            "testsourcequeueDAC45857",
            "Arn",
          ],
        },
        "SourceParameters": Object {
          "FilterCriteria": Object {
            "Filters": Array [
              Object {
                "Pattern": "{\\"foo\\":\\"bar\\"}",
              },
              Object {
                "Pattern": "{\\"md5OfBody\\":\\"string\\",\\"body\\":{\\"foo\\":\\"bar\\"}}",
              },
            ],
          },
        },
        "Target": Object {
          "Fn::GetAtt": Array [
            "testtargetqueue960B8186",
            "Arn",
          ],
        },
        "TargetParameters": Object {
          "SqsQueueParameters": Object {},
        },
      },
      "Type": "AWS::Pipes::Pipe",
    },
    "testpipeRole4CA84261": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "pipes.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testpipeRoleDefaultPolicyC6480FD6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testsourcequeueDAC45857",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testtargetqueue960B8186",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testpipeRoleDefaultPolicyC6480FD6",
        "Roles": Array [
          Object {
            "Ref": "testpipeRole4CA84261",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testsourcequeueDAC45857": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testtargetqueue960B8186": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
